server/matchmaker/algorithm/team_matchmaker.py

```diff
@@ -0,0 +1,262 @@
+import statistics as stats
+from collections import defaultdict
+from typing import Dict, Iterable, List, Tuple
+
+from ...config import config
+from ...decorators import with_logger
+from ..search import CombinedSearch, Game, Match, Search
+from .matchmaker import Matchmaker
+
+
+class Container:
+    def __init__(self, rating_difference, content):
+        self.rating = rating_difference
+        self.content = content
+
+
+@with_logger
+class TeamMatchMaker(Matchmaker):
+    """
+    Matchmaker for teams of varied size. Untested for higher than 4v4 but it should work
+
+    Overview of the algorithm:
+    1. list all the parties in queue by their average rating.
+    2. Take a party and select the neighboring parties alternating between lower
+    and higher until you have 8 players. If the next party to select would leave
+    you with more than 8 players you skip that party and try the next one.
+    3. you now have a list of parties that will be in your potential game. Distribute
+     them into two teams.
+     Start with the easiest cases: one party makes a full team already or one party
+     has n-1 players so you only need to find the best fitting single player.
+     If that is not the case perform the karmarkar-karp algorithm to get a good approximation for a partition
+    4. add this game to a games list
+    5. repeat 2. to 4. for every party.
+    6. you now have a list of potential games with minimal rating variation and minimal rating imbalance.
+    7. remove all games with match quality below threshold then sort by quality descending
+    8. pick the first game from the game list and remove all other games that contain the same players
+    9. repeat until the list is empty
+
+    for the future:
+    Find combination of potential games that allows for maximal number of games to launch
+    Repeat 8. with an increasing amount of top games removed beforehand to get all possible game combinations
+    If there is more than one solution, pick the one with highest total game quality
+    Optimization: sort by game quality first, abort when you find a solution with
+    the full number of theoretically possible games (floor(playersInQueue/(teamSize*2)))
+    """
+
+    def find(self, searches: Iterable[Search]) -> List[Match]:
+        searches = list(searches)
+        searches.sort(key=lambda s: s.average_rating, reverse=True)
+        self._logger.debug("=== starting matching algorithm ===")
+        possible_games = set()
+        for index, search in enumerate(searches):
+            self._logger.debug("building game for %s", repr(search))
+            participants = self._pick_neighboring_players(searches, index)
+            try:
+                match = self.make_teams(list(participants))
```

Comment:
Copying list for no reason. IMO it should be the callee's responsibility to make a copy if it needs to do that to prevent modifying arguments.
```suggestion
                match = self.make_teams(participants)
```

Reply:
In this particular case actually, the function *could* take ownership of the participants list and modify it directly, because I don't see it being used anywhere else. Probably better not to do that though.

Reply:
So if I just delete the copy in the call like you suggested will that make the function take ownership of the list like you said in your last comment, or does it need extra steps to do that?
I.e. should I just delete the copy in the call and leave the function as is, or should I make additional changes in the function?

Reply:
Well sadly there isn't a concept of ownership built into the language, as everything in python is a shared reference. I mean "Taking ownership" in a conceptual sense. It would just mean you make sure you don't use the argument after you've passed it to the function because you know that the function will assume that it owns that data and modify it. You would have to document the behavior clearly in the docstring.

Something like this for example:
```python3
takes_foo(makes_foo())
```
The `takes_foo` function can do whatever it wants to the output from `makes_foo` because it's never assigned to a variable so it can't even be used in the outer scope.